filament/backend/src/JobQueue.cpp

```diff
@@ -0,0 +1,229 @@
+    size_t jobsToTake = mJobOrder.size();
+    if (0 < maxJobsToPop && maxJobsToPop < static_cast<int>(jobsToTake)) {
+        jobsToTake = maxJobsToPop;
+    }
+    jobs.reserve(jobsToTake);
+
+    while (0 < jobsToTake && !mJobOrder.empty()) {
+        JobId jobId = mJobOrder.front();
+        mJobOrder.pop();
+
+        auto it = mJobsMap.find(jobId);
+        if (UTILS_UNLIKELY(it == mJobsMap.end())) {
+            // The job was probably canceled.
+            continue;
+        }
+
+        jobs.push_back(std::move(it->second));
+        --jobsToTake;
+        mJobsMap.erase(it);
+    }
+
+    return jobs;
+}
+
+JobQueue::JobId JobQueue::issueJobId() noexcept {
+    std::lock_guard<std::mutex> lock(mQueueMutex);
+    JobId const jobId = mNextJobId++;
+    // Preallocate a job, which servers two main purposes. It provides a valid jobId that can be
+    // checked for integrity when passed to the `push` method, and it enables job cancellation for
+    // tasks that are yet to be pushed.
+    mJobsMap[jobId];
+    return jobId;
+}
+
+bool JobQueue::cancel(JobId const jobId) noexcept {
+    std::lock_guard<std::mutex> lock(mQueueMutex);
+
+    auto it = mJobsMap.find(jobId);
+    if (it == mJobsMap.end()) {
+        return false; // Job not found, must have been completed or canceled.
+    }
+
+    mJobsMap.erase(it);
+
+    return true;
+}
+
+void JobQueue::stop() noexcept {
+    {
+        std::lock_guard<std::mutex> lock(mQueueMutex);
+        mIsStopping = true;
+    }
+    mQueueCondition.notify_all(); // Wake up all waiting threads
+}
+
+JobWorker::~JobWorker() = default;
+
+void JobWorker::terminate() {
+    if (mQueue) {
+        mQueue->stop();
+    }
+}
+
+AmortizationWorker::AmortizationWorker(JobQueue::Ptr queue, PassKey)
+    : JobWorker(std::move(queue)) {
+}
+
+void AmortizationWorker::process(int const jobCount) {
+    if (!mQueue || jobCount == 0) {
+        return;
+    }
+
+    if (jobCount == 1) {
+        // Handle single job without vector allocation.
+        if (auto job = mQueue->pop(false)) {
+            job();
+        }
+        return;
+    }
+
+    // Handle batch (jobCount > 1 or jobCount < 0 for "all pending jobs")
+    utils::FixedCapacityVector<JobQueue::Job> jobs = mQueue->popBatch(jobCount);
+    if (jobs.empty()) {
+        return;
+    }
+
+    for (auto& job: jobs) {
+        job();
+    }
+}
+
+void AmortizationWorker::terminate() {
+    JobWorker::terminate();
+
+    // Drain all pending jobs.
+    process(-1);
+}
+
+ThreadWorker::ThreadWorker(JobQueue::Ptr queue, Config config, PassKey)
+        : JobWorker(std::move(queue)), mConfig(std::move(config)) {
+    mThread = std::thread([this]() {
+        utils::JobSystem::setThreadName(mConfig.name.data());
+        utils::JobSystem::setThreadPriority(mConfig.priority);
+
+        if (mConfig.onBegin) {
+            mConfig.onBegin();
+        }
+
+        while (JobQueue::Job job = mQueue->pop(true)) {
+            job();
+        }
```

Comment:
So the worker just exits if/when the queue is empty? It doesn't wait for more jobs?